DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp

```diff
@@ -0,0 +1,552 @@
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+#include "single_kernel.hpp"
+#include "multi_kernel.hpp"
+
+using namespace sycl;
+using namespace std::chrono;
+
+// data types and constants
+// NOTE: this tutorial assumes you are using a sycl::vec datatype. Therefore, 
+// 'Type' can only be changed to a different vector datatype (e.g. int16,
+// ulong8, etc...)
+using Type = long8;
+
+///////////////////////////////////////////////////////////////////////////////
+// forward declaration of the functions in this file
+// the function definitions are all below the main() function in this file
+template<typename T>
+void DoWorkOffload(queue& q, T* in, T* out, size_t total_count,
+                   size_t iterations);
+
+template<typename T>
+void DoWorkSingleKernel(queue& q, T* in, T* out,
+                        size_t chunks, size_t chunk_count, size_t total_count,
+                        size_t inflight_kernels, size_t iterations);
+
+template <typename T>
+void DoWorkMultiKernel(queue& q, T* in, T* out,
+                       size_t chunks, size_t chunk_count, size_t total_count,
+                       size_t inflight_kernels, size_t iterations);
+
+template<typename T>
+void PrintPerformanceInfo(std::string print_prefix, size_t count,
+                          std::vector<double>& latency_ms,
+                          std::vector<double>& process_time_ms);
+///////////////////////////////////////////////////////////////////////////////
+
+
+int main(int argc, char* argv[]) {
+  // default values
+#if defined(FPGA_EMULATOR)
+  size_t chunks = 1 << 4;         // 16
+  size_t chunk_count = 1 << 8;    // 256
+  size_t iterations = 2;
+#else
+  size_t chunks = 1 << 9;         // 512
+  size_t chunk_count = 1 << 15;   // 32768
+  size_t iterations = 5;
+#endif
+
+  // This is the number of kernels we will have in the queue at a single time.
+  // If this number is set too low (e.g. 1) then we don't take advantage of
+  // fast kernel relaunch (see the README). If this number is set to high,
+  // then the first kernel launched finishes before we are done launching all
+  // the kernels and therefore throughput is decreased.
+  size_t inflight_kernels = 2;
+
+  // parse the command line arguments
+  for (int i = 1; i < argc; i++) {
+    std::string arg(argv[i]);
+
+    if (arg == "--help" || arg == "-h") {
+      std::cout << "USAGE: "
+                << "./simple_host_streaming "
+                << "[--chunks=<int>] "
+                << "[--chunk_count=<int>] "
+                << "[--inflight_kernels=<int>] "
+                << "[--iterations=<int>]\n";
+      return 0;
+    } else {
+      std::string str_after_equals = arg.substr(arg.find("=") + 1);
+
+      if (arg.find("--chunks=") == 0) {
+        chunks = atoi(str_after_equals.c_str());
```

Comment:
```suggestion
        chunks = std::stoi(str_after_equals);
```
everywhere

Reply:
Don't really feel like wrapping a try-catch around this.

Reply:
I see... I missed the
```c++
  // check the chunks
  if (chunks <= 0) {
    std::cerr << "ERROR: 'chunks' must be greater than 0\n";
    return 1;
  }
```
later.

But honestly a try catch is not that bad
```c++
try { chunks = std::stoi(str_after_equals); }
catch (...) {
  std::cerr << "ERROR: 'chunks' must be greater than 0\n";
  throw; // or std::terminate()
}
```